src/Integration/Aws/SimpleS3/src/SimpleS3Client.php

```diff
@@ -21,8 +28,69 @@ public function getUrl(string $bucket, string $key): string
         return $this->getEndpoint($uri, [], null);
     }
 
-    /*
-     * TODO Add get() and put() which supports large files
-     * TODO add has()
+    public function download(string $bucket, string $key): ResultStream
+    {
+        return $this->getObject(['Bucket' => $bucket, 'Key' => $key])->getBody();
+    }
+
+    public function has(string $bucket, string $key): bool
+    {
+        return $this->objectExists(['Bucket' => $bucket, 'Key' => $key])->isSuccess();
+    }
+
+    /**
+     * @param string|resource|callable|iterable $object
      */
+    public function upload(string $bucket, string $key, $object): void
+    {
+        // Send upload requests with size 64 MB
+        $chunkSize = 64 * 1024 * 1024;
+        $directUpload = \is_string($object);
+        if (!$directUpload && \is_resource($object)) {
```

Comment:
what's about creating stream first and use the `Stream::size` method then ?

Reply:
Because if the $object is an iterable or a callable, then we reading things into memory. This is for objects measured in gigabytes.

Reply:
hmm, no it shouldn't. if yes it's a bug.
the stream return a size when available, otherwise return null

Reply:
You are correct. Thank you

Reply:
note: streams have been designed to not store thing in memory (even hash method)